wgsl-struct-alignment-proposal.md

```diff
@@ -132,6 +132,43 @@ fn foo() -> vec2<f32> {
 }
 ```
 
+#### Interaction with atomics
+
+Atomics are coming. See https://github.com/gpuweb/gpuweb/issues/1360
+
+The widen-to-vec4 technique doesn't work directly if one of the fields being absorbed is atomic:
+```rust
+[[block]] struct Uniforms {
+    a : vec2<u32>;
+    b : atomic<u32>;
+    c : u32;
+};
+```
+
+Suppose we translate it like this:
+```rust
+[[block]] struct Uniforms {
+    a : vec4<u32>;
+};
+```
+
+Consider the Vulkan translation. 
+We should only touch the `b` component via atomic operations.
+We might introduce data races if we do non-atomic loads and stores on the whole vector or on what would be the `b` component. 
```

Comment:
> We might introduce data races if we do non-atomic loads and stores on the whole vector or on what would be the `b` component. 

It's not clear to me how non-atomic loads could be an issue here. As I understand the atomics proposal, they only operate on scalars, and non-atomic vector loads have no guarantees w.r.t atomic stores, right?

As for stores:

Given the apply-swizzles-to-address-scalars-in-the-vec4 proposal above, when do you foresee whole vector stores? Are you concerned that `foo.vec.y = x` might optimize down to:
Insert `x` into a local copy of `foo.vec`, then store the whole vector to the storage of `foo.vec`?

If so - yes, that's a risk - but also one with allowing any sort of vector - atomic packing. In which case it sounds like we need a rule to forbid atomics being packed with any vectors, not just in this compatibility transform.

Reply:
Yes, the danger is doing normal loads and stores on the whole vector. That introduces new normal loads and stores that can race with atomics from the "other" thread.

Original program:

       Thread 0:
              atomic store to foo.b
              normal store to foo.a
             

        Thread 1:
              atomic load from foo.b

This has no race because the only locations touched by both threads is the atomic b and it's only via atomic operations.


Translated program with widening:

       Thread 0:
              atomic store to foo.b
              normal store to vec4 covering all 16 bytes: a b c
             

        Thread 1:
              atomic load from foo.b

Now we have a normal store to b in one thread and an atomic load from the other thread.   T

his doesn't occur in C++ because of type rules and that atomics are separate types.   

SPIR-V Vulkan allows you to intermix normal and atomic accesses provided you synchronize properly.  But if all of thread 0 executes then all of thread 1 executes then we have a race from the second store in thread 0 to the atomic load in thread 1.

If I swap loads for stores and vice versa I still get a race in a different schedule but for the other reason:  The store from thread 1 can race the vector load in thread 0.

I agree with the conclusion:  you can't do a vector load or store when it covers an atomic.  

We can either ban the mixing, or break everything down to scalars.

Reply:
I'll update the PR, now that I was forced to have good examples.  :-)